internal/querynodev2/segments/segment.go

```diff
@@ -579,6 +584,65 @@ func (s *LocalSegment) Retrieve(ctx context.Context, plan *RetrievePlan) (*segco
 	return result, nil
 }
 
+func (s *LocalSegment) RetrieveByOffsets(ctx context.Context, plan *RetrievePlan, offsets []int64) (*segcorepb.RetrieveResults, error) {
+	if !s.ptrLock.RLockIf(state.IsNotReleased) {
+		// TODO: check if the segment is readable but not released. too many related logic need to be refactor.
+		return nil, merr.WrapErrSegmentNotLoaded(s.ID(), "segment released")
+	}
+	defer s.ptrLock.RUnlock()
+
+	if s.ptr == nil {
+		return nil, merr.WrapErrSegmentNotLoaded(s.ID(), "segment released")
+	}
+
+	fields := []zap.Field{
+		zap.Int64("collectionID", s.Collection()),
+		zap.Int64("partitionID", s.Partition()),
+		zap.Int64("segmentID", s.ID()),
+		zap.Int64("msgID", plan.msgID),
+		zap.String("segmentType", s.segmentType.String()),
+		zap.Int("resultNum", len(offsets)),
+	}
+
+	log := log.Ctx(ctx).With(fields...)
+	log.Debug("begin to retrieve by offsets")
+
+	traceCtx := ParseCTraceContext(ctx)
+
+	var retrieveResult RetrieveResult
+	var status C.CStatus
+	GetSQPool().Submit(func() (any, error) {
+		tr := timerecord.NewTimeRecorder("cgoRetrieveByOffsets")
+		status = C.RetrieveByOffsets(traceCtx,
+			s.ptr,
+			plan.cRetrievePlan,
+			&retrieveResult.cRetrieveResult,
+			(*C.int64_t)(unsafe.Pointer(&offsets[0])),
```

Comment:
maybe need to check the length of `offsets` to prevent panic

Reply:
OK.